Make a C++11 compiler mandatory if INDI or C++ binding are requested

dforsi opened the pull request:
Otherwise make it optional.

Fixes issue #1730.

Regardless if a C++ compiler is installed or not, nothing changes when running `configure` without arguments or with explicit `--with-cxx-binding --with-indi`.

If a C++ compiler is NOT installed, before this PR the configure script fails even if called as:
> ./configure --without-cxx-binding --without-indi
```
configure: error: *** A compiler with support for C++11 language features is required.
```
We can simulate the absence of a C++ compiler passing `CXX=/usr/bin/false` to configure:
> ./configure  CXX=/usr/bin/false --without-indi --without-cxx-binding 

With this PR, configure succeeds and prints
```
    With C++ binding                no
...
    With INDI support               no
```
The build succeeds.

N0NB commented:
Seems the test not seeing `x"$cf_with_cxx"` with the `"xyes"` value.  On Linux I sww that `config.log` has `cf_with_cxx='yes'`.  I'll have to do some checking to see if the `config.log` file can be read via the CI UI.

dforsi commented:
Yes, reading `config.log` would save time.

In this case the issue is caused by the line

> AC_CHECK_PROG([cf_with_cxx], ["${CXX}"], [yes], [no])

I added and `echo ${CXX}` on Mac is `g++ -std=gnu++11` not just the name of the program.

Comparing the output showed that `-std=gnu++11` was a difference:

Ubuntu
> checking for gcc option to enable C11 features... (cached) none needed
> ...
> checking for g++... yes

Mac
> checking for g++ option to enable C++11 features... -std=gnu++11
> ...
> checking for g++ -std=gnu++11... no

### Ubuntu
```
checking dependency style of gcc... gcc3
checking whether make supports nested variables... (cached) yes
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking for g++... g++
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... none needed
checking dependency style of g++... gcc3
checking how to run the C preprocessor... gcc -E
checking for gawk... (cached) gawk
checking whether ln -s works... yes
checking for g++... yes
```

### Mac
```
checking dependency style of gcc... gcc3
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking for g++... g++
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... -std=gnu++11
checking dependency style of g++ -std=gnu++11... gcc3
checking how to run the C preprocessor... gcc -E
checking for gawk... (cached) awk
checking whether ln -s works... yes
checking for g++ -std=gnu++11... no
```

markjfine commented:
I'm thinking this might be a clang thing masquerading as something gnu++11-compatible. Lemme pull this down and see what I end up with. I have gcc installed via Homebrew, but I don't think I have gnu-c++.

markjfine commented:
Mine won't build with INDI because it fails the libnova test. I don't have that installed.

Otherwise, here's what I'm getting with no configure switches:
```
checking dependency style of gcc... gcc3
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking for g++... g++
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... none needed
checking dependency style of g++... gcc3
checking how to run the C preprocessor... gcc -E
checking for gawk... (cached) awk
checking whether ln -s works... yes
*** CXX is 'g++'
checking for g++... yes
```
the output shows:
```
 Hamlib Version 4.7~git configuration:

 Prefix 	/usr/local
 Preprocessor	gcc -E 
 C Compiler	gcc -g -O2 
 C++ Compiler	g++ -std=c++11 -g -O2

 Package features:
    With C++ binding		    yes
    With Perl binding		    no
    With Python binding 	    no
    With TCL binding		    no
    With Lua binding		    no
    With rigmem XML support	    no
    With Readline support	    yes
    With INDI support		    no

    Enable HTML rig feature matrix  no
    Enable WinRadio		    yes
    Enable Parallel		    yes
    Enable USRP 		    no
    Enable USB backends 	    no
    Enable shared libs		    yes
    Enable static libs		    yes
    Enable Python tests		    no
```
Note that it says `c++11`, not `gnu++11`

if I add CXXFLAGS=-std=gnu++11 it says the same during configuration, but it just adds the `-std=gnu++11` to the C++ Compiler line in the output:
```
C++ Compiler	g++ -std=c++11 -std=gnu++11
```

Again, I'm thinking this may be a clang thing.
Let me know if there's anything specific you need me to look at and I'll be more than happy to oblige.

dforsi commented:
Yes, please @markjfine can you confirm that passing `false` as C++ compiler is a valid test case also for your system?

Running `./configure CXX=/usr/bin/false --without-indi --without-cxx-binding` should finish without errors, show that the C++ compiler is "/usr/bin/false" and that C++ binding and INDI are both "no".

It may be a clang thing because it needs the extra argument, but the error is that `AC_CHECK_PROG()` expects the name of a file instead it receives an entire command line on github: it gets `g++ -std=gnu++11` instead of `g++` and it tries to find a file named exactly `g++ -std=gnu++11`

markjfine commented:
`./configure CXX=/usr/bin/false --without-indi --without-cxx-binding`
Completes with no errors:
```
Hamlib Version 4.7~git configuration:

 Prefix 	/usr/local
 Preprocessor	gcc -E 
 C Compiler	gcc -g -O2 
 C++ Compiler	/usr/bin/false 

 Package features:
    With C++ binding		    no
    With Perl binding		    no
    With Python binding 	    no
    With TCL binding		    no
    With Lua binding		    no
    With rigmem XML support	    no
    With Readline support	    yes
    With INDI support		    no

    Enable HTML rig feature matrix  no
    Enable WinRadio		    yes
    Enable Parallel		    yes
    Enable USRP 		    
    Enable USB backends 	    no
    Enable shared libs		    yes
    Enable static libs		    yes
    Enable Python tests		    no
```

markjfine commented:
I should add this is an M1 running macOS 15.6.1 (just updated this morning), using Apple's xCode 16.4 as well as Homebrew's gcc v15.1.0.

N0NB commented:
Thanks, Mark!

markjfine commented:
You're both certainly welcome.

N0NB commented:
I see a typo in your last commit.  `dml` should be `dnl` on line 114.

It seems to me that if a C++ compiler is not found that the script should disable INDI support and continue on without error.  For example:

```
$ ../hamlib/configure CXX=/usr/bin/false
```

Should not error out as follows:

```
checking whether to use INDI in rigctl/rotctl... yes
configure: error: INDI support needs a C++ compiler.
```

I don't even have the Debian `libindi-dev` package installed and apparently never have had it installed.  Evidently it not being available is fine so long as the C++ compiler is found, but if there is no such compiler then it fails with a hard error.  Can we make it handle this case gracefully?

dforsi commented:
It was the goal of this PR to fail the configuration step if you run `./configure --with-indi --with-cxx-binding` (which is the default) and the prerequisites aren't met.

Why "yes" should be ignored?

We can change the default to `detect` or `auto` or some other word that we decide (it's just a string that must be checked to be different from "yes" and "no"):
* `detect` is the same as when no arguments are passed (the behavior is like before, ie build if the prerequisites are met, print a warning otherwise)
* `no` do not build (the behavior is like before)
* `yes` build if prerequisites are met, fail otherwise (the new behavior)

dforsi commented:
I added the option `detect` to `--with-cxx-binding` and `--with-indi` and it works as said: `yes` causes a fatal error if something isn't available, `no` disables building the option and `detect` builds it if it finds the prerequisites.

I removed the old `AC_MSG_WARN()` because yes/no is printed anyway during the checks and it makes `configure.ac` a little simpler:

```
checking whether to use INDI in rotctl/rotctld... detect
checking for nova library... checking for libnova/libnova.h... yes
-lnova -lstdc++ -lz
checking for INDI library... checking for libindi/baseclient.h... yes
-lindiclient -lstdc++ -lz
```
I think that the line with `-lnova -lstdc++ -lz` is printed by the macro `AX_LIB_NOVA` and `-lindiclient -lstdc++ -lz` by `AX_LIB_INDI`, I don't know why they weren't visible before.

I used the following script to test with the C++ compiler enabled/disabled and libindi installed/uninstalled:
```
for cxx in "" "CXX=false"; do
    for options in \
        "" \
        "--with-cxx-binding=yes --with-indi=yes" \
        "--with-cxx-binding=no --with-indi=no";
    do
        echo ./configure $cxx $options
        ./configure $cxx $options 2>&1 | grep -iE "( indi | nova |c\+\+ binding)"
        echo
    done
done | tee output.txt
```
[output.txt](https://github.com/user-attachments/files/21957159/output.txt)
[output-no-libindi.txt](https://github.com/user-attachments/files/21957160/output-no-libindi.txt)

I also installed libnova in the CI for macOS and added `--with-indi=detect`, it doesn't find it (a path issue on Mac?) so the configure continues disabling INDI.

dforsi commented:
I think that's a good idea because we can see if a failure is only in one OS.
It will also make it easier to add another workflow for Windows or
Raspberry or something else.
In the current workflow I like the sharing of the `--with` and
`--enable` options, while overriding values for some steps, I don't
know how to share them between different scripts.

I see that workflows can be executed manually with the trigger `on
workflow_dispatch`, this can be useful to test the split scripts
before having them run automatically, there isn't much documentation
because it seems to work similar to `on push`/`on pull_request`:

https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/
https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax?s=09#onworkflow_dispatch
